dipu/torch_dipu/csrc_dipu/binding/ExportTensor.cpp

```diff
@@ -35,28 +38,31 @@ static at::Tensor dispatch_to(
       non_blocking, copy);
 }
 
-// TODO(fandaoyi): check memory leak
 // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-static std::shared_ptr<PyObject* [2]> splitArgs(PyObject* args) {
+static auto splitArgs(PyObject* args) {
   ssize_t rawSize = PyTuple_Size(args);
-  PyObject* newArgs = PyTuple_New(rawSize - 1);
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
-  std::shared_ptr<PyObject* [2]> result(new PyObject*[2], [](PyObject** p) {
-    // if (p[1]) {    // cause segfault, why?
-    //   Py_DECREF(p[1]);
-    // }
-    delete[] p;
-    p = nullptr;
-  });
-  // 0 is self
-  result[0] = PyTuple_GET_ITEM(args, 0);
-  result[1] = newArgs;
-
+  PyObject* newArgsTuple = PyTuple_New(rawSize - 1);
   for (int i = 1; i < rawSize; i++) {
     auto arg = PyTuple_GET_ITEM(args, i);
-    PyTuple_SetItem(newArgs, i - 1, arg);
+    Py_INCREF(arg);
+    PyTuple_SetItem(newArgsTuple, i - 1, arg);
   }
-  return result;
+
+  auto makeDeleter = [](bool isContainer) {
+    return [isContainer](PyObject* p) {
+      if (isContainer) {
+        Py_DecRef(p);
+      }
+    };
+  };
+
+  auto selfDeleter = makeDeleter(false);
+  auto newArgsDeleter = makeDeleter(true);
+  auto self = unique_ptr<PyObject, decltype(selfDeleter)>(
+      PyTuple_GET_ITEM(args, 0), selfDeleter);
+  auto newArgs = unique_ptr<PyObject, decltype(newArgsDeleter)>(newArgsTuple,
+                                                                newArgsDeleter);
```

Comment:
写绕了吧，既然也没归并为一个类型，不如直接写两个 lambda 清楚
另外，无捕获的 lambda 可以转为函数指针
```suggestion
  using PyObjectUniquePointer = std::unique_ptr<PyObject, void(*)(PyObject*)>
  auto self = PyObjectUniquePointer(PyTuple_GET_ITEM(args, 0), [](PyObject*) {});
  auto newArgs = PyObjectUniquePointer(newArgsTuple, [](PyObject* p) { Py_DecRef(p); });
```

Reply:
fix